lib/kube/proxy/responsewriters/watcher.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package responsewriters
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/streaming"
+	"k8s.io/apimachinery/pkg/watch"
+	restclientwatch "k8s.io/client-go/rest/watch"
+)
+
+const (
+	ContentTypeHeader  = "Content-Type"
+	DefaultContentType = "application/json"
+)
+
+// WatcherResponseWriter satisfies the http.ResponseWriter interface and
+// once the server writes the headers and response code spins a go routine
+// that parses each event frame, decodes it and analyzes if the user
+// is allowed to receive the events for that pod.
+// If the user is not allowed, the event is ignored.
+// If allowed, the event is encoded into the user's response.
+type WatcherResponseWriter struct {
```

Comment:
I would rather test this type on its own too, if possible, but up to you.

Reply:
It is tested in the `proxy` package via `TestWatcherResponseWriter` to simplify the required code.